core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala

```diff
@@ -0,0 +1,207 @@
+    webUi.bind()
+    scheduler.frameworkUrl = webUi.activeWebUiUrl
+    scheduler.start()
+    server.start()
+  }
+
+  def awaitShutdown(): Unit = {
+    shutdownLatch.await()
+  }
+
+  def stop(): Unit = {
+    webUi.stop()
+    server.stop()
+    scheduler.stop()
+    shutdownLatch.countDown()
+  }
+}
+
+private[mesos] object MesosClusterDispatcher extends spark.Logging {
+  def main(args: Array[String]) {
+    SignalLogger.register(log)
+
+    val conf = new SparkConf
+    val dispatcherArgs = new ClusterDispatcherArguments(args, conf)
+
+    conf.setMaster(dispatcherArgs.masterUrl)
+    conf.setAppName("Spark Cluster")
+
+    dispatcherArgs.zookeeperUrl.foreach { z =>
+      conf.set("spark.deploy.recoveryMode", "ZOOKEEPER")
+      conf.set("spark.deploy.zookeeper.url", z)
+    }
+
+    val dispatcher = new MesosClusterDispatcher(
+      dispatcherArgs,
+      conf)
+
+    dispatcher.start()
+
+    val shutdownHook = new Thread() {
+      override def run() {
+        logInfo("Shutdown hook is shutting down dispatcher")
+        dispatcher.stop()
+        dispatcher.awaitShutdown()
+      }
+    }
+
+    Runtime.getRuntime.addShutdownHook(shutdownHook)
+
+    dispatcher.awaitShutdown()
+  }
+
+  private class ClusterDispatcherArguments(args: Array[String], conf: SparkConf) {
+    var host = Utils.localHostName()
+    var port = 7077
+    var webUiPort = 8081
+    var masterUrl: String = _
+    var zookeeperUrl: Option[String] = None
+    var propertiesFile: String = _
+
+    parse(args.toList)
+
+    propertiesFile = Utils.loadDefaultSparkProperties(conf, propertiesFile)
+
+    private def parse(args: List[String]): Unit = args match {
+      case ("--host" | "-h") :: value :: tail =>
+        Utils.checkHost(value, "Please use hostname " + value)
+        host = value
+        parse(tail)
+
+      case ("--port" | "-p") :: IntParam(value) :: tail =>
+        port = value
+        parse(tail)
+
+      case ("--webui-port" | "-p") :: IntParam(value) :: tail =>
+        webUiPort = value
+        parse(tail)
+
+      case ("--zk" | "-z") :: value :: tail =>
+        zookeeperUrl = Some(value)
+        parse(tail)
+
+      case ("--master" | "-m") :: value :: tail =>
+        if (!value.startsWith("mesos://")) {
+          System.err.println("Cluster dispatcher only supports mesos (uri begins with mesos://)")
+          System.exit(1)
+        }
+        masterUrl = value.stripPrefix("mesos://")
+        parse(tail)
+
+      case ("--properties-file") :: value :: tail =>
+        propertiesFile = value
+        parse(tail)
+
+      case ("--help") :: tail =>
+        printUsageAndExit(0)
+
+      case Nil => {
+        if (masterUrl == null) {
+          System.err.println("--master is required")
+          System.exit(1)
+        }
+      }
+
+      case _ =>
+        printUsageAndExit(1)
+    }
+
+    /**
+     * Print usage and exit JVM with the given exit code.
+     */
+    def printUsageAndExit(exitCode: Int): Unit = {
```

Comment:
can be private I believe. Also the javadoc is somewhat unnecessary since it doesn't provide any extra information than the signature itself